### ISSUE TITLE
[SingleMulti-TS] Loading Similarities

### ISSUE THREAD

**Issue by khairunnisaor** [microsoft/vert-papers#9]
Hi,

Thanks for your great work on this project.

I got a problem when I was running the Multi-source teacher-student learning. I think it's sentence/domain similarity related.
It appears on the `# STEP3: multi-source teacher-student learning` after I run the domain model training and save the domain model.

```2021-06-20 13:53:14 WARNING: - device: cuda, n_gpu: 1
2021-06-20 13:53:14 INFO: - Training/evaluation parameters Namespace(adam_epsilon=1e-08, balance_classes=False, cache_dir='', config_name='', data_dir='./data/ner/conll', device=device(type='cuda'), do_lower_case=False, do_predict=False, do_train=True, do_voting=False, domain_orthogonal=False, evaluate_during_training=False, freeze_bottom_layer=3, gamma_R=0.01, gpu_ids=[7], hard_label_usage='none', hard_label_weight=0, labels='./data/ner/conll/labels.txt', learning_rate=0.0001, log_dir='result-22/ts-learn-var-domain-id-rank_64-gamma_0.01/logs', logging_steps=20, low_rank_size=64, max_grad_norm=1.0, max_seq_length=128, max_steps=-1, model_name_or_path='bert-base-multilingual-cased', n_gpu=1, num_train_epochs=3, output_dir='result-22/ts-learn-var-domain-id-rank_64-gamma_0.01', overwrite_cache=False, per_gpu_eval_batch_size=64, per_gpu_train_batch_size=32, save_steps=2000, seed=22, sim_dir='domain-model', sim_level='domain', sim_type='learn', sim_with_tgt=False, src_langs=['en'], src_model_dir='conll-model-22', src_model_dir_prefix='mono-src-', tau_metric='var', tgt_lang='id', tokenizer_name='', train_hard_label=False, warmup_ratio=0.1, warmup_steps=0, weight_decay=1e-05)
2021-06-20 13:53:16 INFO: - Loading features from cached file ./data/ner/conll/en/cached_train_bert-base-multilingual-cased_128.
2021-06-20 13:53:18 INFO: - ********** scheme: training with KD **********
2021-06-20 13:53:18 INFO: - Loading features from cached file ./data/ner/conll/id/cached_train_bert-base-multilingual-cased_128.
2021-06-20 13:53:18 INFO: - sim_dir: domain-model/id-rank_64-gamma_0.01-seed_22
2021-06-20 13:53:18 INFO: - ==> Computing similarities....
2021-06-20 13:53:18 INFO: - ***** Compute sentence embeddings for [id] plain text dataset using the [pre-trained] base_model *****
2021-06-20 13:53:29 INFO: -   Num examples = 1
2021-06-20 13:53:29 INFO: -   Batch size = 64
2021-06-20 13:53:29 INFO: - ***** Running evaluation on id *****
2021-06-20 13:53:29 INFO: -   Num of examples = 1
2021-06-20 13:53:29 INFO: -   Instantaneous batch size per GPU = 64
2021-06-20 13:53:29 INFO: -   GPU IDs for training: 7
2021-06-20 13:53:29 INFO: -   Batch size = 64
2021-06-20 13:53:29 INFO: - ==> tau: nan
2021-06-20 13:53:29 INFO: -   Sentence similarities:
2021-06-20 13:53:29 INFO: -   en
2021-06-20 13:53:29 INFO: -   nan
Traceback (most recent call last):
  File "main.py", line 918, in <module>
    main(args)
  File "main.py", line 822, in main
    src_probs, src_predictions = get_src_weighted_probs(args, dataset_pt, config, mode="train", src_datasets=src_datasets)
  File "main.py", line 590, in get_src_weighted_probs
    st_sims, dm_sims = get_st_sims(args, domain_model, dataset_st_ebd, src_idxs)
  File "main.py", line 489, in get_st_sims
    print_sim_info(st_sims, dm_sims)
  File "main.py", line 536, in print_sim_info
    logger.info("  " + "\t".join([str(round(v.item(), 4)) for v in st_sims[i]]))
IndexError: index 2 is out of bounds for dimension 0 with size 1
```

Do you have any idea why this happens and how to fix this?

Thanks in advance!

**Comment by QianhuiWu** [comment]
> Hi,
> 
> Thanks for your great work on this project.
> 
> I got a problem when I was running the Multi-source teacher-student learning. I think it's sentence/domain similarity related.
> It appears on the `# STEP3: multi-source teacher-student learning` after I run the domain model training and save the domain model.
> 
> ```
> 2021-06-20 13:53:14 INFO: - Training/evaluation parameters Namespace(adam_epsilon=1e-08, balance_classes=False, cache_dir='', config_name='', data_dir='./data/ner/conll', device=device(type='cuda'), do_lower_case=False, do_predict=False, do_train=True, do_voting=False, domain_orthogonal=False, evaluate_during_training=False, freeze_bottom_layer=3, gamma_R=0.01, gpu_ids=[7], hard_label_usage='none', hard_label_weight=0, labels='./data/ner/conll/labels.txt', learning_rate=0.0001, log_dir='result-22/ts-learn-var-domain-id-rank_64-gamma_0.01/logs', logging_steps=20, low_rank_size=64, max_grad_norm=1.0, max_seq_length=128, max_steps=-1, model_name_or_path='bert-base-multilingual-cased', n_gpu=1, num_train_epochs=3, output_dir='result-22/ts-learn-var-domain-id-rank_64-gamma_0.01', overwrite_cache=False, per_gpu_eval_batch_size=64, per_gpu_train_batch_size=32, save_steps=2000, seed=22, sim_dir='domain-model', sim_level='domain', sim_type='learn', sim_with_tgt=False, src_langs=['en'], src_model_dir='conll-model-22', src_model_dir_prefix='mono-src-', tau_metric='var', tgt_lang='id', tokenizer_name='', train_hard_label=False, warmup_ratio=0.1, warmup_steps=0, weight_decay=1e-05)
> 2021-06-20 13:53:16 INFO: - Loading features from cached file ./data/ner/conll/en/cached_train_bert-base-multilingual-cased_128.
> 2021-06-20 13:53:18 INFO: - ********** scheme: training with KD **********
> 2021-06-20 13:53:18 INFO: - Loading features from cached file ./data/ner/conll/id/cached_train_bert-base-multilingual-cased_128.
> 2021-06-20 13:53:18 INFO: - sim_dir: domain-model/id-rank_64-gamma_0.01-seed_22
> 2021-06-20 13:53:18 INFO: - ==> Computing similarities....
> 2021-06-20 13:53:18 INFO: - ***** Compute sentence embeddings for [id] plain text dataset using the [pre-trained] base_model *****
> 2021-06-20 13:53:29 INFO: -   Num examples = 1
> 2021-06-20 13:53:29 INFO: -   Batch size = 64
> 2021-06-20 13:53:29 INFO: - ***** Running evaluation on id *****
> 2021-06-20 13:53:29 INFO: -   Num of examples = 1
> 2021-06-20 13:53:29 INFO: -   Instantaneous batch size per GPU = 64
> 2021-06-20 13:53:29 INFO: -   GPU IDs for training: 7
> 2021-06-20 13:53:29 INFO: -   Batch size = 64
> 2021-06-20 13:53:29 INFO: - ==> tau: nan
> 2021-06-20 13:53:29 INFO: -   Sentence similarities:
> 2021-06-20 13:53:29 INFO: -   en
> 2021-06-20 13:53:29 INFO: -   nan
> Traceback (most recent call last):
>   File "main.py", line 918, in <module>
>     main(args)
>   File "main.py", line 822, in main
>     src_probs, src_predictions = get_src_weighted_probs(args, dataset_pt, config, mode="train", src_datasets=src_datasets)
>   File "main.py", line 590, in get_src_weighted_probs
>     st_sims, dm_sims = get_st_sims(args, domain_model, dataset_st_ebd, src_idxs)
>   File "main.py", line 489, in get_st_sims
>     print_sim_info(st_sims, dm_sims)
>   File "main.py", line 536, in print_sim_info
>     logger.info("  " + "\t".join([str(round(v.item(), 4)) for v in st_sims[i]]))
> IndexError: index 2 is out of bounds for dimension 0 with size 1
> ```
> 
> Do you have any idea why this happens and how to fix this?
> 
> Thanks in advance!

Hi, what are the source languages in your experiment? When you run main.py, it is supposed that there were more than one source language, and the language similarities should be computed with domain_learner.py in advance.

From the log, it seems that you only have one source language now? If so, please try to run the single-teacher setting (main_single.py)~

**Comment by tellarin** [comment]
@khairunnisaor I'll close this issue now. Feel free to continue the conversation here if you have additional question on this.

**Comment by khairunnisaor** [comment]
Hi, thanks for your replies.

I want to use English, Spanish, and Dutch as sources to be transferred to Indonesian. And yes, I've run the domain_learner.py (since I run the `run_multi.sh` which is contained the domain_learner, but then I got this error. Do you have any idea about this?

Thanks!

**Comment by QianhuiWu** [comment]
> Hi, thanks for your replies.
> 
> I want to use English, Spanish, and Dutch as sources to be transferred to Indonesian. And yes, I've run the domain_learner.py (since I run the `run_multi.sh` which is contained the domain_learner, but then I got this error. Do you have any idea about this?
> 
> Thanks!

Hi, thank you for your interest in our paper. There're two issues you may check. 

First, from the 15th row of the shown log, it seems that the variable \tau was not calculated correctly. Is there only one example in your target language dataset? If so, please use more examples and then have a try~

Second, from the 17th row of the shown log, it seems that only English is recognized as the source language. For this issue, maybe you can check the 7th line of `run_multi.sh` to make sure that multiple source languages are required.

Hope it helps!

**Comment by khairunnisaor** [comment]
Hi, thanks for your prompt reply.

I've put the en, es, and nl dataset in a separate folder as shown in the sample inside `./data/ner/conll/` folder and got the `conll-model-${seed}/mono-src-${lan}` results. Then, I noticed that in the Step 2

```# STEP2: train domain model
            python domain_learner.py \
                --do_train \
                --gpu_ids ${GPU_IDS} \
                --seed ${seed} \
                --data_dir ./data/ner/conll \
                --labels ./data/ner/conll/labels.txt \
                --src_langs ${LANGS} \
                --tgt_lang ${TGT_LANG} \
                --gamma_R ${g} \
                --low_rank_size ${rank} \
                --tau_metric var \
                --num_train_epochs 10 \
                --output_dir domain-model/${TGT_LANG}-rank_${rank}-gamma_${g}-seed_${seed}```

the data_dir is `./data/ner/conll` (following what you gave in the `run_multi.sh`). Did I put the multi-lang data correctly for the domain learner?

Thank you so much!

**Comment by QianhuiWu** [comment]
> Hi, thanks for your prompt reply.
> 
> I've put the en, es, and nl dataset in a separate folder as shown in the sample inside `./data/ner/conll/` folder and got the `conll-model-${seed}/mono-src-${lan}` results. Then, I noticed that in the Step 2
> 
> ```
>             # STEP2: train domain model
>             python domain_learner.py \
>                 --do_train \
>                 --gpu_ids ${GPU_IDS} \
>                 --seed ${seed} \
>                 --data_dir ./data/ner/conll \
>                 --labels ./data/ner/conll/labels.txt \
>                 --src_langs ${LANGS} \
>                 --tgt_lang ${TGT_LANG} \
>                 --gamma_R ${g} \
>                 --low_rank_size ${rank} \
>                 --tau_metric var \
>                 --num_train_epochs 10 \
>                 --output_dir domain-model/${TGT_LANG}-rank_${rank}-gamma_${g}-seed_${seed}
> ```
> 
> the data_dir is `./data/ner/conll` (following what you gave in the `run_multi.sh`). Did I put the multi-lang data correctly for the domain learner?
> 
> Thank you so much!

What's the setting of `LANGS` and `TGT_LANG` in your `run_multi.sh`? How many examples in your Indonesian dataset?

For example, if you want to take English (`./data/ner/conll/en`) , Spanish (`./data/ner/conll/es`), and Dutch (`./data/ner/conll/nl`) as sources and regard Indonesian (`./data/ner/conll/id`) as the target language, please make sure that `LANGS=(en es nl)` and `TGT_LANG=id`. 

Also, make sure that there are more than one examples in `./data/ner/conll/id`, otherwise \tau cannot be calculated correctly under the `var` tau_metric.

**Comment by khairunnisaor** [comment]
Hi,

Yes, I already set `LANGS=(en es nl)` and `TGT_LANG=id`. I also include the train, dev, and test file separately for each language. In case of the Indonesian, it has about 2K sentences, but the EN, ES, and NL are more than 10K sentences each (the CoNLL dataset). Does the dataset size difference matters here?

Thanks in advance!

**Comment by QianhuiWu** [comment]
> Hi,
> 
> Yes, I already set `LANGS=(en es nl)` and `TGT_LANG=id`. I also include the train, dev, and test file separately for each language. In case of the Indonesian, it has about 2K sentences, but the EN, ES, and NL are more than 10K sentences each (the CoNLL dataset). Does the dataset size difference matters here?
> 
> Thanks in advance!

I checked that information because I noticed that the experiment log provided by you shows that `Num examples = 1` (which means the dataset size parsed into the function is 1, but it should be around 2K according to your reply). 

Also, only `en` was recognized as the source language, but the corresponding codes are:
``` python
logger.info("  Sentence similarities:")
logger.info("  " + "\t".join(args.src_langs))
```
which should be `en es nl`. Maybe something is wrong with the source-language arguments (`args.src_langs`).

To sum up, I think maybe we can make sure that the data format is consistent with the shown examples, and check why `args.src_langs` are not parsed the way we want.

**Comment by khairunnisaor** [comment]
Hi,

Thank you for your thorough explanation.

According to that, maybe I was wrong in giving the data format. 
As I understand, I've followed the example in your code but I will check it again, maybe character separator difference or other things. I will update here if I found it and make it works.

Thank you so much!
